rust/src/pgsql/parser.rs

```diff
@@ -0,0 +1,2088 @@
+    pub data_type_size: i16,
+    // "The value will generally be -1 for types that do not need pg_attribute.atttypmod."
+    pub type_modifier: i32,
+    // "The format code being used for the field. Currently will be zero (text) or one (binary). In a RowDescription returned from the variant of Describe, will always be zero"
+    pub format_code: u16,
+}
+
+// TODO should I have a constructor here?
+#[derive(Debug, PartialEq)]
+pub struct RowDescriptionMessage {
+    pub identifier: u8,
+    pub length: u32,
+    pub field_count: u16,
+    pub fields: Vec<RowField>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ColumnFieldValue {
+    // Can be 0, or -1 as a special NULL column value
+    pub value_length: i32,
+    // If length is -1, no value bytes will follow
+    // value will have `value_length` bytes
+    // TODO - get rid of "option"?
+    pub value: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DataRowMessage {
+    pub identifier: u8,
+    pub length: u32,
+    pub field_count: u16,
+    pub fields: Vec<ColumnFieldValue>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum PgsqlErrorNoticeFieldType {
+    SeverityLocalizable,
+    SeverityNonLocalizable,
+    CodeSqlStateCode,
+    Message,
+    Detail,
+    Hint,
+    Position,
+    InternalPosition,
+    InternalQuery,
+    Where,
+    SchemaName,
+    TableName,
+    ColumnName,
+    DataType,
+    ConstraintName,
+    File,
+    Line,
+    Routine,
+    TerminatorToken,
+    UnknownFieldType,
+}
+
+impl From<char> for PgsqlErrorNoticeFieldType {
+    fn from(identifier: char) -> PgsqlErrorNoticeFieldType {
+        match identifier {
+            'S' => PgsqlErrorNoticeFieldType::SeverityLocalizable,
+            'V' => PgsqlErrorNoticeFieldType::SeverityNonLocalizable,
+            'C' => PgsqlErrorNoticeFieldType::CodeSqlStateCode,
+            'M' => PgsqlErrorNoticeFieldType::Message,
+            'D' => PgsqlErrorNoticeFieldType::Detail,
+            'H' => PgsqlErrorNoticeFieldType::Hint,
+            'P' => PgsqlErrorNoticeFieldType::Position,
+            'p' => PgsqlErrorNoticeFieldType::InternalPosition,
+            'q' => PgsqlErrorNoticeFieldType::InternalQuery,
+            'W' => PgsqlErrorNoticeFieldType::Where,
+            's' => PgsqlErrorNoticeFieldType::SchemaName,
+            't' => PgsqlErrorNoticeFieldType::TableName,
+            'c' => PgsqlErrorNoticeFieldType::ColumnName,
+            'd' => PgsqlErrorNoticeFieldType::DataType,
+            'n' => PgsqlErrorNoticeFieldType::ConstraintName,
+            'F' => PgsqlErrorNoticeFieldType::File,
+            'L' => PgsqlErrorNoticeFieldType::Line,
+            'R' => PgsqlErrorNoticeFieldType::Routine,
+            '\u{0}' => PgsqlErrorNoticeFieldType::TerminatorToken,
+            _ => PgsqlErrorNoticeFieldType::UnknownFieldType, // adding this because documentation says "frontends should silently ignore fields of unrecognized type."
+        }
+    }
+}
+
+impl From<u8> for PgsqlErrorNoticeFieldType {
```

Comment:
can these 2 funcs somehow be merged, or use a common util func?

Reply:
Checking this properly, now, those two cannot, as this `From<u8>` is for the `ErrorMessages` and `NoticeMessages` types, not the front end messages.

The From functions are also used by the nom parsers, in the cases where that makes sense, and return types compatible with said parsers. Whereas the get_message_type is used by the app-layer parsing the requests and responses. So I'm guessing that some documentation is needed here, to explain that (plus proper naming the `get_message_type`one)...